trl/trainer/utils.py

```diff
@@ -358,6 +358,224 @@ def __call__(self, features: List[Dict[str, Any]]) -> Dict[str, Any]:
 
 
+@dataclass
+class SPINDataCollatorWithPadding:
+    r"""
+    DataCollator class that pads the inputs to the maximum length of the batch.
+    Args:
+        tokenizer (`PreTrainedTokenizerBase`):
+            The tokenizer used for encoding the data.
+        model (Optional[`PreTrainedModel`]):
+            The model that is being trained. If set and has the *prepare_decoder_input_ids_from_labels*, use it to
+            prepare the *decoder_input_ids*.
+        padding (`Union[bool, str, `PaddingStrategy`]`, `optional`, defaults to `True`):
+            padding_strategy to pass to the tokenizer.
+        max_length (`Optional[int]`, `optional`, defaults to `None`):
+            The maximum length of the sequence to be processed.
+        max_prompt_length (`Optional[int]`, `optional`, defaults to `None`):
+            The maximum length of the prompt to be processed.
+        label_pad_token_id (`int`, defaults to -100):
+            The label used for masking.
+        padding_value (`int`, defaults to 0):
+            The value used for padding.
+        is_encoder_decoder (`Optional[bool]`, `optional`, defaults to `None`):
+            Whether or not you model has an encoder_decoder architecture.
+        max_target_length (`Optional[int]`, `optional`, defaults to `None`):
+            The maximum length of the target to be processed. Only useful for encoder-decoder architectures.
+        truncation_mode: (`str`, defaults to "keep_end"):
+            The truncation mode to use when truncating the prompt.
+    """
+    tokenizer: PreTrainedTokenizerBase
+    model: Optional[PreTrainedModel] = None
+    padding: Union[bool, str] = True
+    max_length: Optional[int] = None
+    max_prompt_length: Optional[int] = None
+    label_pad_token_id: int = -100
+    padding_value: int = 0
+    truncation_mode: str = "keep_end"
+    is_encoder_decoder: Optional[bool] = False
+    max_target_length: Optional[int] = None
+
+    def tokenize_batch_element(
+        self,
+        prompt: str,
+        chosen: str,
+        rejected: str,
+    ) -> Dict:
+        """Tokenize a single batch element.
+
+        At this stage, we don't convert to PyTorch tensors yet; we just handle the truncation
+            in case the prompt + chosen or prompt + rejected responses is/are too long. First
+            we truncate the prompt; if we're still too long, we truncate the chosen/rejected.
+
+        We also create the labels for the chosen/rejected responses, which are of length equal to
+            the sum of the length of the prompt and the chosen/rejected response, with
+            label_pad_token_id  for the prompt tokens.
+        """
+        batch = {}
+
+        if not self.is_encoder_decoder:
+            chosen_tokens = self.tokenizer(chosen, add_special_tokens=False)
+            rejected_tokens = self.tokenizer(rejected, add_special_tokens=False)
+            prompt_tokens = self.tokenizer(prompt, add_special_tokens=False)
```

Comment:
This feels quite similar to the logic in DPOTrainer. Maybe unify the logic somehow (or use similar terminology). 

https://github.com/huggingface/trl/blob/b32656f7266c02500ea346eda75a0db7a9a1bb75/trl/trainer/dpo_trainer.py#L721-L729

Not necessary but we could also benefit from doing a filediff between SPIN and DPO and try to make the lines of code differences minimal.

<img width="1597" alt="image" src="https://github.com/huggingface/trl/assets/5555347/d69ab518-af3a-4b7c-a317-8dd736efc4d7">